### PR TITLE
Storefront product sort

### PR DESCRIPTION
**Pull request by szymoniwacz**
1. update storefront products sort specs
2. enable products sort by available_on

### CODE REVIEW

File: api/docs/v2/storefront/index.yaml [hunk 1]
```diff
@@ -1208,13 +1208,16 @@ paths:
           name: sort
           schema:
             type: string
-          example: '-updated_at,price,-name'
+          example: '-updated_at,price,-name,created_at,-available_on,variant.sku'
```
Comment (damianlegawiec) [inline api/docs/v2/storefront/index.yaml:1211]:
```suggestion
          example: '-updated_at,price,-name,created_at,-available_on,sku'
```

File: api/docs/v2/storefront/index.yaml [hunk 2]
```diff
@@ -1208,13 +1208,16 @@ paths:
...
-              <li>price (ascending/descenging)</li>
-              <li>updated_at (ascending/descenging)</li>
-              <li>name (ascending/descenging)</li>
-            </ul> Use <q>-</q> sign to set descenging sort, eg. <q>-updated_at</q>
+              <li>price (ascending/descending)</li>
+              <li>updated_at (ascending/descending)</li>
+              <li>created_at (ascending/descending)</li>
+              <li>name (ascending/descending)</li>
+              <li>available_on (ascending/descending)</li>
+              <li>variant.sku (ascending/descending)</li>
```
Comment (damianlegawiec) [inline api/docs/v2/storefront/index.yaml:1219]:
```suggestion
              <li>sku (ascending/descending)</li>
```

File: core/app/sorters/spree/products/sort.rb [hunk 3]
```diff
@@ -9,6 +9,7 @@ def initialize(scope, current_currency, params = {}, allowed_sort_attributes = [
       def call
         products = by_param_attribute(scope)
         products = by_price(products)
+        products = by_variant_sku(products)
```
Comment (damianlegawiec) [inline core/app/sorters/spree/products/sort.rb:12]:
```suggestion
        products = by_sku(products)
```

File: core/app/sorters/spree/products/sort.rb [hunk 4]
```diff
@@ -30,6 +31,19 @@ def by_price(scope)
           where(spree_prices: { currency: currency }).
           order("#{Spree::Price.table_name}.amount #{order_direction}")
       end
+
+      def variant_sku?
+        sort_field == 'variant.sku'
```
Comment (damianlegawiec) [inline core/app/sorters/spree/products/sort.rb:36]:
```suggestion
      def sku?
        sort_field == 'sku'
```

File: core/app/sorters/spree/products/sort.rb [hunk 5]
```diff
@@ -30,6 +31,19 @@ def by_price(scope)
...
+      end
+
+      def by_variant_sku(scope)
+        return scope unless variant_sku?
+
+        scope.joins(:master).
+          select("#{Spree::Product.table_name}.*, #{Spree::Variant.table_name}.sku").
```
Comment (damianlegawiec) [inline core/app/sorters/spree/products/sort.rb:43]: @szymoniwacz can we avoid this `select(`?

**Review by damianlegawiec — COMMENTED** [review]
@szymoniwacz let's break this PR into 2, the SKU thing is more complex

**Review by damianlegawiec — APPROVED** [review]

File: api/app/controllers/spree/api/v2/storefront/products_controller.rb [hunk 6]
```diff
@@ -42,6 +42,10 @@ def model_class
           def scope_includes
             product_list_includes
           end
+
+          def allowed_sort_attributes
```
Comment (damianlegawiec) [inline api/app/controllers/spree/api/v2/storefront/products_controller.rb:46]: @szymoniwacz let's also add this change to the platform api as well - https://github.com/spree/spree/blob/master/api/app/controllers/spree/api/v2/platform/products_controller.rb

**Review by damianlegawiec — APPROVED** [review]

**Review by v10110 — APPROVED** [review]

**Pull request merged**